networkx/algorithms/community/louvain.py

```diff
@@ -0,0 +1,531 @@
+
+    Parameters
+    ----------
+    dendogram : list of dict
+       a list of partitions, ie dictionnaries where keys of the i+1 are the values of the i.
+    level : int
+       the level which belongs to [0..len(dendogram)-1]
+
+    Returns
+    -------
+    partition : dictionnary
+       A dictionary where keys are the nodes and the values are the set it belongs to
+
+    Raises
+    ------
+    KeyError
+       If the dendogram is not well formed or the level is too high
+
+    See Also
+    --------
+    best_partition which directly combines partition_at_level and generate_dendogram to obtain the partition of highest modularity
+
+    Examples
+    --------
+    >>> G=nx.erdos_renyi_graph(100, 0.01)
+    >>> dendo = generate_dendogram(G)
+    >>> for level in range(len(dendo) - 1) :
+    >>>     print "partition at level", level, "is", partition_at_level(dendo, level)
+    """
+    partition = dendogram[0].copy()
+    for index in range(1, level + 1) :
+        for node, community in partition.items() :
+            partition[node] = dendogram[index][community]
+    return partition
+
+
+def modularity(partition, graph) :
+    """Compute the modularity of a partition of a graph
+
+    Parameters
+    ----------
+    partition : dict
+       the partition of the nodes, i.e a dictionary where keys are their nodes and values the communities
+    graph : networkx.Graph
+       the networkx graph which is decomposed
+
+    Returns
+    -------
+    modularity : float
+       The modularity
+
+    Raises
+    ------
+    KeyError
+       If the partition is not a partition of all graph nodes
+    ValueError
+        If the graph has no link
+    TypeError
+        If graph is not a networkx.Graph
+
+    References
+    ----------
+    .. 1. Newman, M.E.J. & Girvan, M. Finding and evaluating community structure in networks. Physical Review E 69, 26113(2004).
+
+    Examples
+    --------
+    >>> G=nx.erdos_renyi_graph(100, 0.01)
+    >>> part = best_partition(G)
+    >>> modularity(part, G)
+    """
+    if type(graph) != nx.Graph :
+        raise TypeError("Bad graph type, use only non directed graph")
+
+    inc = dict([])
+    deg = dict([])
+    links = graph.size(weight='weight')
+    if links == 0 :
+        raise ValueError("A graph without link has an undefined modularity")
+
+    for node in graph :
+        com = partition[node]
+        deg[com] = deg.get(com, 0.) + graph.degree(node, weight = 'weight')
+        for neighbor, datas in graph[node].items() :
+            weight = datas.get("weight", 1)
+            if partition[neighbor] == com :
+                if neighbor == node :
+                    inc[com] = inc.get(com, 0.) + float(weight)
+                else :
+                    inc[com] = inc.get(com, 0.) + float(weight) / 2.
+
+    res = 0.
+    for com in set(partition.values()) :
+        res += (inc.get(com, 0.) / links) - (deg.get(com, 0.) / (2.*links))**2
+    return res
```

Comment:
Simpler: `return sum(inc[c] / links - (deg[c] / (2 * links)) ** 2 for c in set(partition.values()))`.